onnxruntime/core/graph/function_utils.cc

```diff
@@ -114,17 +114,29 @@ static void IOTypeConstraintHelper(const ONNX_NAMESPACE::FunctionProto& onnx_fun
 
   std::function<void(const ONNX_NAMESPACE::NodeProto&)> process_node = [&](const ONNX_NAMESPACE::NodeProto& node) {
     auto it = opset_imports.find(node.domain());
-    ORT_ENFORCE(it != opset_imports.end(), "No opset registered for domain " + node.domain() + " in function opset imports.");
+    ORT_ENFORCE(it != opset_imports.end(),
+                "No opset registered for domain " + node.domain() + " in function opset imports.");
     int domain_version = it->second;
-    ORT_ENFORCE(domain_version != -1, "No opset registered for domain " + node.domain() + " in function opset imports.");
-    const auto node_op_schema =
-        schema_registry->GetSchema(node.op_type(), domain_version, node.domain());
+    ORT_ENFORCE(domain_version != -1,
+                "No opset registered for domain " + node.domain() + " in function opset imports.");
+
+    const auto* node_op_schema = schema_registry->GetSchema(node.op_type(), domain_version, node.domain());
+    int variadic_arg_idx = -1;
     for (int i = 0; i < node.input_size(); ++i) {
       auto& in_name = node.input().Get(i);
       auto iter = input_name_idx_map.find(in_name);
       if (iter != input_name_idx_map.end()) {
         int idx = iter->second;
-        std::string type_str = node_op_schema ? node_op_schema->inputs().at(i).GetTypeStr() + "in" + std::to_string(idx) : "Tin" + std::to_string(idx);
+        // if we have hit a variadic arg it is the last input in the schema, so we need to use that index not i.
+        auto schema_idx = variadic_arg_idx != -1 ? variadic_arg_idx : i;
+
+        std::string type_str;
+        if (node_op_schema) {
+          type_str = node_op_schema->inputs().at(schema_idx).GetTypeStr() + "in" + std::to_string(schema_idx);
+        } else {
+          type_str = "Tin" + std::to_string(idx);
+        }
```

Comment:
Not sure if this is correct. Differs slightly from the original code for both inputs and outputs. 

If the type string is coming from the schema, always use the idx for the schema input/output. 
Otherwise use the idx value from the name to idx map. I am guessing that map is to create one type constraint when there are multiple inputs/outputs with the same name, but am not 100% sure.